velox/expression/tests/ExpressionFuzzerVerifier.h

```diff
@@ -0,0 +1,142 @@
+
+  void reSeed();
+
+  // A utility class used to keep track of stats relevant to the fuzzer.
+  class ExprStatsListener : public exec::ExprSetListener {
+   public:
+    explicit ExprStatsListener(
+        std::unordered_map<std::string, ExprUsageStats>& exprNameToStats)
+        : exprNameToStats_(exprNameToStats) {}
+
+    void onCompletion(
+        const std::string& /*uuid*/,
+        const exec::ExprSetCompletionEvent& event) override {
+      for (auto& [funcName, stats] : event.stats) {
+        auto itr = exprNameToStats_.find(funcName);
+        if (itr == exprNameToStats_.end()) {
+          // Skip expressions like FieldReference and ConstantExpr
+          continue;
+        }
+        itr->second.numProcessedRows += stats.numProcessedRows;
+      }
+    }
+
+    // A no-op since we cannot tie errors directly to functions where they
+    // occurred.
+    void onError(
+        const SelectivityVector& /*rows*/,
+        const ::facebook::velox::ErrorVector& /*errors*/,
+        const std::string& /*queryId*/) override {}
+
+   private:
+    std::unordered_map<std::string, ExprUsageStats>& exprNameToStats_;
+  };
+
+  /// Randomize initial result vector data to test for correct null and data
+  /// setting in functions.
+  RowVectorPtr generateResultVectors(std::vector<core::TypedExprPtr>& plans);
+
+  /// Executes two steps:
+  /// #1. Retries executing the expression in `plan` by wrapping it in a `try()`
+  ///     clause and expecting it not to throw an exception.
+  /// #2. Re-execute the expression only on rows that produced non-NULL values
+  ///     in the previous step.
+  ///
+  /// Throws in case any of these steps fail.
+  void retryWithTry(
+      std::vector<core::TypedExprPtr> plans,
+      const RowVectorPtr& rowVector,
+      const VectorPtr& resultVectors,
+      const std::vector<int>& columnsToWrapInLazy);
+
+  /// If --duration_sec > 0, check if we expired the time budget. Otherwise,
+  /// check if we expired the number of iterations (--steps).
+  template <typename T>
+  bool isDone(size_t i, T startTime) const;
+
+  /// Called at the end of a successful fuzzer run. It logs the top and bottom
+  /// 10 functions based on the num of rows processed by them. Also logs a full
+  /// list of all functions sorted in descending order by the num of times they
+  /// were selected by the fuzzer. Every logged function contains the
+  /// information in the following format which can be easily exported to a
+  /// spreadsheet for further analysis: functionName numTimesSelected
+  /// proportionOfTimesSelected numProcessedRows.
+  void logStats();
+
+  FuzzerGenerator rng_;
+
+  size_t currentSeed_{0};
+
+  std::shared_ptr<core::QueryCtx> queryCtx_{std::make_shared<core::QueryCtx>()};
+
+  std::shared_ptr<memory::MemoryPool> pool_{memory::addDefaultLeafMemoryPool()};
+
+  core::ExecCtx execCtx_{pool_.get(), queryCtx_.get()};
+
+  test::ExpressionVerifier verifier_;
+
+  std::shared_ptr<VectorFuzzer> vectorFuzzer_;
```

Comment:
If ExpressionFuzzer holds a VectorFuzzer&, could this simply be `VectorFuzzer vectorFuzzer_`?

Reply:
maybe the user does not need to hold into the fuzzer in that case it might be better to be more flexible here